File: jswebsocket/UniRpc/WebsocketService.ts
import { WebsocketServiceBase } from "./WebsocketServiceBase";
import { IRequestContext, IWebSocketConnection, IWebsocketEvent, IWebSocketUser } from "./LambdaWebsocketTypes";
import { BaseMessage } from "./BaseMessage";
import { GroupClausesAuthorize } from './GroupAuthorizations'
import { ApiGatewayManagementApi, DynamoDB } from 'aws-sdk';

const WebSocketConnectionsTable = process.env['WebSocketConnectionsTable'];
const WebSocketUsersTable = process.env['WebSocketUsersTable'];
const dynamo = new DynamoDB();

export class WebsocketService {
  tracking: {[serial: string]: any} = {};
  services: Map<string, WebsocketServiceBase> = new Map<string, WebsocketServiceBase>();
  user: IWebSocketUser;

  RegisterService<T extends WebsocketServiceBase>(service: T): this {
    this.services.set(service.__reflection, service);
    return this;
  }

  async ProcessEvent(event: IWebsocketEvent) {
    try {
      this.user = await this.GetUser(event.requestContext);
    } catch (ex) {
      console.error(ex);
      return {
        statusCode: 401,
        body: 'Unauthorized'
      };
    }
    let message: BaseMessage = JSON.parse(event.body);
    if (!GroupClausesAuthorize(this.user.Groups.S, message.Service, message.Method)) {
      return {
        statusCode: 401,
        body: 'Unauthorized'
      };
    }
    if (this.services.has(message.Service)) {
      let service = this.services.get(message.Service);
      let result = await service.__invoke(message);
      await this.Respond(event.requestContext, result);
      return {
        statusCode: 202,
        body: 'Accepted'
      };
    }
    return {
      statusCode: 403,
      body: 'Forbidden'
    };
  }

  async GetUser(context: IRequestContext): Promise<IWebSocketUser> {
    let connectionResponse = await (dynamo.getItem({
        TableName: WebSocketConnectionsTable,
        Key: {
            ConnectionId: { S: context.connectionId }
        }
    }).promise());
    if(!connectionResponse.Item) throw `No Connection was found for Connection Id ${context.connectionId}`;
    let connection: IWebSocketConnection = connectionResponse.Item as any;
    let userResponse = await (dynamo.getItem({
        TableName: WebSocketUsersTable,
        Key: {
            Username: { S: connection.Username.S }
        }
    }).promise());
    if(!userResponse.Item) throw `No user was found for User Id ${connection.Username.S} via Connection Id ${context.connectionId}`;
    return userResponse.Item as any;
  }

  async RespondUnauthorized(event: IWebsocketEvent, message: BaseMessage) {
    let response: BaseMessage = {} as any;
    response.Id = message.Id;
    response.Service = message.Service;
    response.Method = message.Method;
    response.Success = false;
    response.ErrorMessage = `Unauthorzied`;
    await this.Respond(event.requestContext, response);
  }

  async Respond(context: IRequestContext, data: any): Promise<void> {
    let agm = new ApiGatewayManagementApi({
        endpoint: `${context.domainName}/${context.stage}`
    });
    let stringData = JSON.stringify(data);
    await (agm.postToConnection({
        ConnectionId: context.connectionId,
        Data: stringData
    }).promise());
  }
}

File: tests/transformers/python/websocket-lambda-service/WebsocketService.py
from __future__ import annotations
from typing import Dict, Any
import json
import boto3
import boto3_type_annotations.dynamodb
import boto3_type_annotations.apigatewaymanagementapi
import os

from WebsocketServiceBase import WebsocketServiceBase
from LambdaWebsocketTypes import IWebSocketUser, IWebsocketEvent, IRequestContext, IWebSocketConnection
from BaseMessage import BaseMessage
from GroupAuthorizations import GroupClausesAuthorize


WebSocketConnectionsTable = os.getenv('WebSocketConnectionsTable')
WebSocketUsersTable = os.getenv('WebSocketUsersTable')
dynamo: boto3_type_annotations.dynamodb.Client = boto3.client('dynamodb')

class WebsocketService:
    tracking: Dict[str, Any] = dict()
    services: Dict[str, WebsocketServiceBase] = dict()
    user: IWebSocketUser

    def RegisterService(self, service: WebsocketServiceBase) -> WebsocketService:
        self.services.set(service['__reflection'], service)
        return self

    async def ProcessEvent(self, event: IWebsocketEvent):
        try:
            self.user = await self.GetUser(event.requestContext)
        except Exception as exception:
            print(exception)
            return {
                'statusCode': 401,
                'body': 'Unauthorized'
            }
        message: BaseMessage = json.loads(event.body)
        if not GroupClausesAuthorize(self.user.Groups.S, message.Service, message.Method):
            return {
                'statusCode': 401,
                'body': 'Unauthorized'
            }
        if message.Service in self.services.keys():
            service = self.services.get(message.Service)
            result = await service.__invoke(message)
            self.Respond(event.requestContext, result);
            return {
                'statusCode': 202,
                'body': 'Accepted'
            }
        return {
            'statusCode': 403,
            'body': 'Forbidden'
        }

    async def GetUser(self, context: IRequestContext):
        connectionId = context['connectionId']
        connectionResponse: Dict[str, Any] = dynamo.get_item({
            'TableName': WebSocketConnectionsTable,
            'Key': {
                'ConnectionId': {
                    'S': connectionId
                }
            }
        })
        if 'Item' not in connectionResponse.keys():
            raise Exception(f'No Connection was found for Connection Id {connectionId}')
        connection: IWebSocketConnection = connectionResponse['Item']
        username = connection['Username']['S']
        userResponse = dynamo.getItem({
            'TableName': WebSocketUsersTable,
            'Key': {
                'Username': {
                    'S': username
                }
            }
        })
        if 'Item' not in userResponse.keys():
            raise Exception(f'No user was found for User Id {username} via Connection Id ${connectionId}')
        return userResponse['Item']

    def RespondUnauthorized(self, event: IWebsocketEvent, message: BaseMessage):
        response: BaseMessage = {
            'Id': message['Id'],
            'Service': message['Service'],
            'Method': message['Method'],
            'Success': False,
            'ErrorMessage': 'Unauthorzied'
        }
        self.Respond(event.requestContext, response);

    def Respond(self, context: IRequestContext, data: any):
        agm: boto3_type_annotations.apigatewaymanagementapi.Client = boto3.client(
            'apigatewaymanagementapi', endpoint_url=context['domainName'] + '/' + context['stage']
        )
        connectionId: str = context['connectionId']
        stringData: str = json.dumps(data)
        agm.postToConnection({
            'ConnectionId': connectionId,
            'Data': stringData
        })
